src/envoy/auth/controller.cc

```diff
@@ -0,0 +1,175 @@
+/* Copyright 2017 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "controller.h"
+
+namespace Envoy {
+namespace Http {
+namespace Auth {
+namespace {
+
+// The authorization HTTP header.
+const LowerCaseString kAuthorizationKey("authorization");
+
+// The autorization bearer prefix.
+const std::string kBearerPrefix = "Bearer ";
+
+// The HTTP header to pass verified token payload.
+const LowerCaseString kJwtPayloadKey("sec-istio-auth-userinfo");
+
+// The per-request JWT authentication object.
+class AuthRequest : public Logger::Loggable<Logger::Id::http>,
+                    public std::enable_shared_from_this<AuthRequest> {
+ public:
+  AuthRequest(HttpGetFunc http_get_func, PubkeyCache& pubkey_cache,
+              HeaderMap& headers, Controller::DoneFunc on_done)
+      : http_get_func_(http_get_func),
+        pubkey_cache_(pubkey_cache),
+        headers_(headers),
+        on_done_(on_done) {}
+
+  // Verify a JWT token.
+  CancelFunc Verify() {
+    const HeaderEntry* entry = headers_.get(kAuthorizationKey);
+    if (!entry) {
+      // TODO: excludes some health checking paths
+      return DoneWithStatus(Status::JWT_MISSED);
+    }
+
+    // Extract token from header.
+    const HeaderString& value = entry->value();
+    if (value.size() <= kBearerPrefix.length() ||
+        strncmp(value.c_str(), kBearerPrefix.c_str(), kBearerPrefix.length()) !=
+            0) {
+      return DoneWithStatus(Status::BEARER_PREFIX_MISMATCH);
+    }
+
+    // Parse JWT token
+    jwt_.reset(new Jwt(value.c_str() + kBearerPrefix.length()));
+    if (jwt_->GetStatus() != Status::OK) {
+      return DoneWithStatus(jwt_->GetStatus());
+    }
+
+    // Check "exp" claim.
+    auto unix_timestamp =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            std::chrono::system_clock::now().time_since_epoch())
+            .count();
+    if (jwt_->Exp() < unix_timestamp) {
+      return DoneWithStatus(Status::JWT_EXPIRED);
+    }
+
+    // Check the issuer is configured or not.
+    auto issuer = pubkey_cache_.LookupByIssuer(jwt_->Iss());
+    if (!issuer) {
+      return DoneWithStatus(Status::JWT_UNKNOWN_ISSUER);
+    }
+
+    // Check if audience is allowed
+    if (!issuer->config().IsAudienceAllowed(jwt_->Aud())) {
+      return DoneWithStatus(Status::AUDIENCE_NOT_ALLOWED);
+    }
+
+    if (issuer->pubkey() && !issuer->Expired()) {
```

Comment:
issuer->pubkey() != nullptr to be clear

Reply:
No, unique_ptr has `operator bool` and it should be used.